### ISSUE TITLE
Crashing when trying to backup ImageBot conversation

### ISSUE THREAD

**Issue by LeoDJ** [tvdstaaij/telegram-history-dump#8]
I'm not quite sure if this is the issue of your awesome backup script, or of the telegram-cli.
It seems like old content from @ImageBot is no longer available and when the backup got to that conversation, both the telegram-cli and the backup script crashed.

Sadly, I couldn't simply add the ImageBot to a blacklist and backup everything else, so I deleted the chat history (of ImageBot) and now it works fine. (But it wasn't very important anyways)

I wouldn't class is as high-priority issue but it would be great if it wouldn't crash just because of some deleted content. ;)

#### telegram-history-dump:
```
I, [2016-02-17T16:10:42.932076 #4652]  INFO -- : Dumping "ImageBot" (range 1-100)
E, [2016-02-17T16:10:43.052428 #4652] ERROR -- : Failed to download media file: no implicit conversion of nil into String
E, [2016-02-17T16:10:43.053209 #4652] ERROR -- : Failed to download media file: Broken pipe
E, [2016-02-17T16:10:43.053626 #4652] ERROR -- : Failed to download media file: Broken pipe
E, [2016-02-17T16:10:43.053914 #4652] ERROR -- : Failed to download media file: Broken pipe
E, [2016-02-17T16:10:43.054132 #4652] ERROR -- : Failed to download media file: Broken pipe
E, [2016-02-17T16:10:43.054244 #4652] ERROR -- : Failed to download media file: Broken pipe
E, [2016-02-17T16:10:43.054463 #4652] ERROR -- : Failed to download media file: Broken pipe
E, [2016-02-17T16:10:43.054590 #4652] ERROR -- : Failed to download media file: Broken pipe
E, [2016-02-17T16:10:43.054679 #4652] ERROR -- : Failed to download media file: Broken pipe
E, [2016-02-17T16:10:43.054788 #4652] ERROR -- : Failed to download media file: Broken pipe
E, [2016-02-17T16:10:43.055270 #4652] ERROR -- : Failed to download media file: Broken pipe
E, [2016-02-17T16:10:43.055480 #4652] ERROR -- : Failed to download media file: Broken pipe
E, [2016-02-17T16:10:43.055955 #4652] ERROR -- : Failed to download media file: Broken pipe
E, [2016-02-17T16:10:43.056277 #4652] ERROR -- : Failed to download media file: Broken pipe
E, [2016-02-17T16:10:43.056553 #4652] ERROR -- : Failed to download media file: Broken pipe
E, [2016-02-17T16:10:43.056802 #4652] ERROR -- : Failed to download media file: Broken pipe
E, [2016-02-17T16:10:43.057216 #4652] ERROR -- : Failed to download media file: Broken pipe
E, [2016-02-17T16:10:43.057532 #4652] ERROR -- : Failed to download media file: Broken pipe
E, [2016-02-17T16:10:43.057897 #4652] ERROR -- : Failed to download media file: Broken pipe
E, [2016-02-17T16:10:43.058162 #4652] ERROR -- : Failed to download media file: Broken pipe
telegram-history-dump.rb:32:in `write': Broken pipe (Errno::EPIPE)
        from telegram-history-dump.rb:32:in `puts'
        from telegram-history-dump.rb:32:in `exec_tg_command'
        from telegram-history-dump.rb:58:in `block in dump_dialog'
        from /usr/lib/ruby/2.1.0/timeout.rb:91:in `block in timeout'
        from /usr/lib/ruby/2.1.0/timeout.rb:35:in `block in catch'
        from /usr/lib/ruby/2.1.0/timeout.rb:35:in `catch'
        from /usr/lib/ruby/2.1.0/timeout.rb:35:in `catch'
        from /usr/lib/ruby/2.1.0/timeout.rb:106:in `timeout'
        from telegram-history-dump.rb:57:in `dump_dialog'
        from telegram-history-dump.rb:193:in `block in <main>'
        from telegram-history-dump.rb:189:in `each'
        from telegram-history-dump.rb:189:in `each_with_index'
        from telegram-history-dump.rb:189:in `<main>'
```

#### telegram-cli:
```
> SIGNAL received
h/opt/telegram-cli/bin/telegram-cli(print_backtrace+0x20)[0x46f7a0]
/opt/telegram-cli/bin/telegram-cli(termination_signal_handler+0x64)[0x46f824]
/lib/x86_64-linux-gnu/libc.so.6(+0x35180)[0x7ff027458180]
/opt/telegram-cli/bin/telegram-cli(tgl_do_load_photo+0xd)[0x49e10d]
/opt/telegram-cli/bin/telegram-cli(interpreter_ex+0x7d6)[0x477c76]
/opt/telegram-cli/bin/telegram-cli[0x46fd05]
/usr/lib/x86_64-linux-gnu/libevent-2.0.so.5(+0x1b2aa)[0x7ff02979e2aa]
/usr/lib/x86_64-linux-gnu/libevent-2.0.so.5(event_base_loop+0x7fc)[0x7ff0297933dc]
/opt/telegram-cli/bin/telegram-cli(net_loop+0xa4)[0x470cb4]
/opt/telegram-cli/bin/telegram-cli(loop+0x195)[0x471f75]
/opt/telegram-cli/bin/telegram-cli(main+0x2c4)[0x46e084]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf5)[0x7ff027444b45]
/opt/telegram-cli/bin/telegram-cli[0x46e180]
```

**Comment by LeoDJ** [comment]
Edit: Also happens when trying to backup a chat with many ImageBot-Images... 
Therefore had to disable media downloading, which I really wanted to do :(

**Comment by tvdstaaij** [comment]
I'm afraid this is a telegram-cli issue. `SIGNAL received` basically means it crashed unexpectedly, `no implicit conversion of nil into String` means my script got no response from telegram-cli to its `load_photo` command but tried to parse it anyway.

You can probably reproduce this issue with just telegram-cli by doing this:
* Create a JSON dump with telegram-history-dump (you probably have one already) without downloading media
* Identify one of the problematic images in the JSON files and note its 'id' property
* Start telegram-cli normally and enter the command `load_photo xxxxxx`, using the message id you found

I recommend filing this as a telegram-cli issue so it can hopefully get a proper fix. In the meantime, as a workaround, you could hack the script a bit to ignore images from ImageBot (assuming those are the only images triggering this crash). Specifically, after [this check](https://github.com/tvdstaaij/telegram-history-dump/blob/v1.0.0/telegram-history-dump.rb#L81), you could add another one like this:
```ruby
return if msg['from'] && msg['from']['username'] == 'ImageBot'
```

**Comment by LeoDJ** [comment]
Thanks, I will try that later

Really cool project btw, keep up the good work :+1:

**Comment by LeoDJ** [comment]
Thanks, that fix worked (almost).
It got way further in downloading media. But at some point it crashed again...
I looked into the last file it created and found the first photo entry without a "file" attribute. And who would have thought, the message came from ImageBot... Only that it wasn't sent directly but forwarded :unamused: 

So, I quickly threw in that piece of code:
```ruby
return if msg['fwd_from'] && msg['fwd_from']['username'] == 'ImageBot'
```
and now it works like a charm :blush: 

Thanks again for the help, now all my data is safe.

**tl;dr:** For anyone who might come by this issue again, here are the two lines and where I added them:
```ruby
def process_media(dialog, msg)
  return unless msg.include?('media')
  return if msg['from'] && msg['from']['username'] == 'ImageBot'
  return if msg['fwd_from'] && msg['fwd_from']['username'] == 'ImageBot'
```

**Comment by tvdstaaij** [comment]
Nice hacking. Too bad you didn't get the ImageBot images, I'm sure you had a strong emotional attachment to them :wink:

**Comment by tvdstaaij** [comment]
@LeoDJ some additional background: I believe the issue you encountered is vysheng/tg#904. I had the same issue recently, probably caused by people sending gifs (which technically aren't actually gifs by the way, Telegram converts them to videos, heh). If you or anyone else runs into this again, try compiling rnhmjoj's fork or applying the workaround patch manually.

**Comment by blaize9** [comment]
@LeoDJ @tvdstaaij 
I have submitted a pull request to [vysheng/tgl/pull#106](https://github.com/vysheng/tgl/pull/106) which would fix the underlying problem with tgl_do_load_photo which is called load_photo in telegram-cli that causes the client to crash when requesting an Empty Message thus fixing telegram-history-dump.

**Comment by LeoDJ** [comment]
Thank you very much.

**Comment by blaize9** [comment]
@LeoDJ I keep on joining a lot of random channels and have yet to re-encounter this issue.
How does the fix work for you?
